src/ucp/dt/dt_iov.c

```diff
@@ -125,3 +126,31 @@ size_t ucp_dt_iov_count_nonempty(const ucp_dt_iov_t *iov, size_t iovcnt)
     }
     return count;
 }
+
+void ucp_dt_iov_memtype_detect(ucp_context_h context, const ucp_dt_iov_t *iov,
+                               size_t iovcnt, const ucp_request_param_t *param,
+                               ucp_memory_info_t *mem_info)
+{
+    ucp_memory_info_t UCS_V_UNUSED check_mem_info;
+    int UCS_V_UNUSED i;
+
+    if (ucs_likely(iovcnt > 0)) {
+        ucp_memory_detect_param(context, iov->buffer, iov->length, param,
+                                mem_info);
+#if ENABLE_ASSERT
+        for (i = 1; i < iovcnt; ++i) {
+            ucp_memory_detect_param(context, iov[i].buffer, iov[i].length,
+                                    param, &check_mem_info);
+            if ((check_mem_info.type != mem_info->type) ||
+                (check_mem_info.sys_dev != mem_info->sys_dev)) {
+                ucs_fatal("inconsistent iov memtypes: iov[%d]=%u-%u "
```

Comment:
maybe we could fail just the one send op, and not fatal error?

Reply:
let's do this, all functions are inline so it could be done in case ENABLE_PARAM_CHECK is set

Reply:
`ucp_dt_iov_memtype_detect` is not inlined,
also have to propogate status up to `ucp_tag_recv_common` and a couple of functions on the stack are not inlined. 

imo it does not worth the effort